plugins/main/bridge/bridge.go

```diff
@@ -358,103 +360,109 @@ func cmdAdd(args *skel.CmdArgs) error {
 		return err
 	}
 
-	// run the IPAM plugin and get back the config to apply
-	r, err := ipam.ExecAdd(n.IPAM.Type, args.StdinData)
-	if err != nil {
-		return err
-	}
+	// Assume L2 interface only
+	result := &current.Result{CNIVersion: cniVersion}
+	result.Interfaces = []*current.Interface{brInterface, hostInterface, containerInterface}
 
-	// release IP in case of failure
-	defer func() {
-		if !success {
-			os.Setenv("CNI_COMMAND", "DEL")
-			ipam.ExecDel(n.IPAM.Type, args.StdinData)
-			os.Setenv("CNI_COMMAND", "ADD")
+	if isLayer3 {
+		// run the IPAM plugin and get back the config to apply
+		r, err := ipam.ExecAdd(n.IPAM.Type, args.StdinData)
+		if err != nil {
+			return err
 		}
-	}()
 
-	// Convert whatever the IPAM result was into the current Result type
-	result, err := current.NewResultFromResult(r)
-	if err != nil {
-		return err
-	}
-
-	if len(result.IPs) == 0 {
-		return errors.New("IPAM plugin returned missing IP config")
-	}
-
-	result.Interfaces = []*current.Interface{brInterface, hostInterface, containerInterface}
-
-	// Gather gateway information for each IP family
-	gwsV4, gwsV6, err := calcGateways(result, n)
-	if err != nil {
-		return err
-	}
+		// release IP in case of failure
+		defer func() {
+			if !success {
+				os.Setenv("CNI_COMMAND", "DEL")
+				ipam.ExecDel(n.IPAM.Type, args.StdinData)
+				os.Setenv("CNI_COMMAND", "ADD")
+			}
+		}()
 
-	// Configure the container hardware address and IP address(es)
-	if err := netns.Do(func(_ ns.NetNS) error {
-		contVeth, err := net.InterfaceByName(args.IfName)
+		// Convert whatever the IPAM result was into the current Result type
+		result, err = current.NewResultFromResult(r)
```

Comment:
You're setting result.interfaces twice. This variable should be renamed to `ipamResult`.

Reply:
I first assume the interface is L2 so I create a L2 result 

```go
// Assume L2 interface only
result := &current.Result{CNIVersion: cniVersion}
result.Interfaces = []*current.Interface{brInterface, hostInterface, containerInterface}
```

But if the IPAM is configured then 

```go
// Convert whatever the IPAM result was into the current Result type
result, err = current.NewResultFromResult(r)
```

So I can use one variable only for the return 

```go
return types.PrintResult(result, cniVersion)
```

And don't need to check again if the interface is a L2 to decide what variable to return 

```go
if  isLayer3 {
 return types.PrintResult(ipamResult, cniVersion)
}

return types.PrintResult(result, cniVersion)
```

Reply:
@squeed this makes sense - do you still want it changed?

Reply:
I guess I'm still a bit uncomfortable with the structure of the code. The fact that the `result` defined in line 364 is *maybe* overwritten strikes me as a source of later bugs. In other words, someone else might manipulate `result`, not understanding that their changes are maybe going to be overwritten.

Could you only define `result` once and update it as needed?